packages/react-native/ReactCommon/react/nativemodule/idlecallbacks/NativeIdleCallbacks.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <react/renderer/runtimescheduler/RuntimeScheduler.h>
+#include <react/renderer/runtimescheduler/RuntimeSchedulerBinding.h>
+#include <chrono>
+#include "IdleTaskRef.h"
+#include "NativeIdleCallbacks.h"
+
+#ifdef RN_DISABLE_OSS_PLUGIN_HEADER
+#include "Plugins.h"
+#endif
+
+std::shared_ptr<facebook::react::TurboModule> NativeIdleCallbacksModuleProvider(
+    std::shared_ptr<facebook::react::CallInvoker> jsInvoker) {
+  return std::make_shared<facebook::react::NativeIdleCallbacks>(
+      std::move(jsInvoker));
+}
+
+namespace facebook::react {
+
+jsi::Function makeTimeRemainingFunction(
+    jsi::Runtime& runtime,
+    std::shared_ptr<RuntimeScheduler> scheduler,
+    RuntimeSchedulerTimePoint deadline) {
+  return jsi::Function::createFromHostFunction(
+      runtime,
+      jsi::PropNameID::forAscii(runtime, "timeRemaining"),
+      0,
+      [&](
+          jsi::Runtime& runtime,
+          const jsi::Value& /* unused */,
+          const jsi::Value* /* unused */,
+          size_t /* unused */) noexcept -> jsi::Value {
+        auto now = scheduler->now();
+
+        if (scheduler->getShouldYield()) {
+          return jsi::Value(runtime, 0.0);
+        }
+
+        double remainingTime = std::chrono::duration_cast<std::chrono::milliseconds>(deadline - now).count();
+        return jsi::Value(runtime, std::max(remainingTime, 0.0));
+      });
+}
+
+NativeIdleCallbacks::NativeIdleCallbacks(std::shared_ptr<CallInvoker> jsInvoker)
+    : NativeIdleCallbacksCxxSpec(std::move(jsInvoker)) {}
+
+CallbackHandle NativeIdleCallbacks::requestIdleCallback(
+    jsi::Runtime& runtime,
+    jsi::Function callback,
+    std::optional<NativeRequestIdleCallbackOptions> options) {
+    using namespace std::chrono_literals;
+
+    auto binding = RuntimeSchedulerBinding::getBinding(runtime);
+    auto scheduler = binding->getRuntimeScheduler();
+
+    double timeoutMs = -1;
+
+    // handle timeout parameter
+    if (options.has_value()) {
+        auto timeout = options.value().timeout;
+        if (timeout.has_value()) {
+          timeoutMs = timeout.value();
+        }
+    }
+  
+    auto userCallback = std::make_shared<jsi::Function>(std::move(callback));
+
+    auto wrappedCallback = jsi::Function::createFromHostFunction(
+        runtime,
+        jsi::PropNameID::forAscii(runtime, "requestIdleCallback$bodyWrapper"),
+        1,
+        [scheduler, cb = std::move(userCallback)](
+            jsi::Runtime& runtime,
+            const jsi::Value& /* unused */,
+            const jsi::Value* args,
+            size_t /* unused */
+        ) noexcept -> jsi::Value {
+            auto didTimeout = args[0].getBool();
+            // Below is a partial solution and does not comply to WHATWG event-loop standards
```

Comment:
Sorry I just re-read the PR summary and I understood. I forgot about this problem when we changed the design. I think it's fine because we check `getShouldYield` which would return `true` if there's any other higher priority task pending execution. It might be good to explain this in the comment.

Reply:
Not sure if that is related (or is) frame alignment, but we have separate and fixed deadlines for each `idleTask` but this might be problematic if there are lot of `idleTask`s this might reduce the performance. 
Unless I misunderstood or missed anything, the deadline should be computed at `RuntimeScheduler` (or `Scheduler::EventPipeConclusion perhaps?), reduced by `requestAnimationFrame` and timers run time and then read from `IdleCallbacks` TurboModule. I postponed this to reduce changes in `RuntimeScheduler` API as the initial design had this right 😅.

Reply:
So the reason why we have a fixed time for all idle callbacks on Web is that we need to yield the execution of JS at regular intervals to process rendering. That's done in the main thread in JS, so this is never an issue for us. In the case that we need to process a user interaction, we'd be scheduling a task with a higher priority in the scheduler, which would make `getShouldYield` return true, so we'd return `0` the next time the current idle callback checks the time remaining, and we'd stop running idle callbacks after that. Essentially, it's the same behavior as on Web but with a different implementation.

Would be good to clarify this behavior in a comment :)